bugfix/17142-venn-sets-comma-issue

jakubSzuminski opened the pull request:
Fixed #17142, Venn diagram did not work when there was a comma in the string in the `series.data.sets` array.

**Example** of the problem: 
`['A,']` and `['B']` = `'A,,B'` `(combination)` which caused issues when `combination.split(',')`

**Solution:**
I join the elements of the array with a string `SPLIT//COMBINATIONS` which is unlikely to occur in user's code, but maybe we would want to join the sets with another unique string.

circleci-mu commented:
### File size comparison
No differences found

circleci-mu commented:
## Visual test results - No difference found

TorsteinHonsi reviewed (changes requested):
Thanks! Please add a test too. Preferably modify an existing venn test so that we don't add to the test processing time.

Besides, there are some `any` casts in the code that you modified - since you're working on this code now and have an understanding, maybe you can take the opportunity to fix the typing?

Not so important, and don't spend too much time on this: Why are we using a string that we concatenate and split, instead of an array? I understand it if it's used for indexing an object, otherwise an array would probably be cleaner.

pawellysy reviewed (approved):
Thanks!

jakubSzuminski commented:
1. I've added a test case for `,` in strings

2. When it comes to the types, I recognized that there are a few stages there:
- 1 - `relations` are created in `VennUtils.processVennData()` which is an array of objects with only `sets` and `value` properties (even though the `VennRelationObject` also had `circle, overlapping, totalOverlap` as required properties)
- 2 - after `VennSeries.layout()` the `circle` is added to each of the `relations` so I decided to make `circle` optional (because it doesn't appear in the first step, the `relations` are "extended")
- 3 - then, the `overlapping` and `totalOverlap` are added in `addOverlapToSets()` so I also decided to make them optional (as they do not appear in the first and second step)
Then, I fixed `any` casting in a few places by adding some `if` conditions.

3. In this case strings (which come from `Array.join(splitter)`) are used for object indexing so I did not change the method for now (but of course I'll do if you decide that we should). 
Also, I've added a `splitter` as a `public static` variable of `VennSeries` so that the user might easily override the `splitter` string to something else if he happens to use the `highcharts-split` in his `series.venn.sets`

jakubSzuminski commented on ts/Series/Venn/VennUtils.ts:
```diff
@@ -91,26 +91,31 @@ function addOverlapToSets(
+                    };
+                }
+
+                mapOfIdToProps[set] = {
+                    totalOverlap: (mapOfIdToProps[set].totalOverlap || 0) +
+                        relation.value,
+                    overlapping: {
+                        ...(mapOfIdToProps[set].overlapping || {}),
+                        [arr[1 - i]]: relation.value
                     }
```
I ran into a strange TypeScript issue here. I've tried to check if `typeof mapOfIdToProps[set].totalOverlap !== 'undefined'` and then add `relation.value` to it, but it did not seem to work as the TypeScript always threw 2532 error - `Object is possibly 'undefined'` (no matter how many conditions did I check that it is not `undefined`).

TorsteinHonsi reviewed (approved):
Okay, thanks for the clean up! Let's keep i like this.
